File: server/fix-facility-ids.js
/**
 * Script to verify and fix facility IDs for all permit monitors
 */

const { initializeDatabase, permitQueries, saveDatabase } = require('./db');
const https = require('https');

// Known correct facility IDs based on investigation
const KNOWN_FACILITIES = {
  'gates of lodore': '250014',  // Dinosaur Green And Yampa River Permits
  'yampa river': '250014',      // Same facility, different division
  'deerlodge park': '250014',   // Same facility, different division
  'san juan river': '234621',   // San Juan River
  'desolation canyon': '233393', // Desolation Gray - Green River Permit
  'grand canyon': '233262',     // Grand Canyon River
};

/**
 * Test if a facility ID is valid
 */
function testFacilityId(facilityId) {
  return new Promise((resolve) => {
    const url = `https://www.recreation.gov/api/permits/${facilityId}`;

    https.get(url, {
      headers: {
        'User-Agent': 'Mozilla/5.0 (Windows NT 10.0; Win64; x64) AppleWebKit/537.36'
      }
    }, (res) => {
      let data = '';

      res.on('data', (chunk) => {
        data += chunk;
      });

      res.on('end', () => {
        if (res.statusCode === 200) {
          try {
            const json = JSON.parse(data);
            if (json.payload) {
              resolve({
                valid: true,
                facilityName: json.payload.facility_name || 'Unknown',
                divisions: json.payload.divisions || {},
                data: json.payload
              });
            } else {
              resolve({ valid: false, error: 'No payload in response' });
            }
          } catch (e) {
            resolve({ valid: false, error: 'Invalid JSON' });
          }
        } else {
          resolve({ valid: false, error: `HTTP ${res.statusCode}` });
        }
      });
    }).on('error', (e) => {
      resolve({ valid: false, error: e.message });
    });
  });
}

/**
 * Search for facility by name
 */
function searchFacilityByName(name) {
  return new Promise((resolve) => {
    const searchUrl = `https://www.recreation.gov/api/search?q=${encodeURIComponent(name)}&entity_type=permit`;

    https.get(searchUrl, {
      headers: {
        'User-Agent': 'Mozilla/5.0 (Windows NT 10.0; Win64; x64) AppleWebKit/537.36'
      }
    }, (res) => {
      let data = '';

      res.on('data', (chunk) => {
        data += chunk;
      });

      res.on('end', () => {
        if (res.statusCode === 200) {
          try {
            const json = JSON.parse(data);
            if (json.results && json.results.length > 0) {
              resolve(json.results[0].entity_id);
            } else {
              resolve(null);
            }
          } catch (e) {
            resolve(null);
          }
        } else {
          resolve(null);
        }
      });
    }).on('error', () => {
      resolve(null);
    });
  });
}

/**
 * Update a permit's facility ID
 */
function updatePermitFacilityId(permitId, userId, newFacilityId) {
  const db = require('./db').db();

  db.run(
    'UPDATE permits SET facility_id = ?, updated_at = CURRENT_TIMESTAMP WHERE id = ? AND user_id = ?',
    [newFacilityId, permitId, userId]
  );

  saveDatabase();
}

/**
 * Sleep helper
 */
function sleep(ms) {
  return new Promise(resolve => setTimeout(resolve, ms));
}

async function main() {
  console.log('='.repeat(70));
  console.log('Facility ID Verification and Fix Script');
  console.log('='.repeat(70));

  // Initialize database
  await initializeDatabase();

  // Get all enabled permits
  const permits = permitQueries.getAllEnabled.all();

  if (permits.length === 0) {
    console.log('\nNo permits found in database!');
    return;
  }

  console.log(`\nFound ${permits.length} permit(s) to check\n`);

  const updatesNeeded = [];

  for (const permit of permits) {
    console.log('\n' + '='.repeat(70));
    console.log(`Permit #${permit.id}: ${permit.name}`);
    console.log(`User: ${permit.email}`);
    console.log(`Current Facility ID: ${permit.facility_id}`);
    console.log('='.repeat(70));

    // FIRST: Check if this matches a known facility
    const nameLower = permit.name.toLowerCase();
    let knownCorrectId = null;

    for (const [knownName, knownId] of Object.entries(KNOWN_FACILITIES)) {
      if (nameLower.includes(knownName)) {
        knownCorrectId = knownId;
        console.log(`\nMatches known facility: "${knownName}" -> ${knownId}`);
        break;
      }
    }

    // If we have a known correct ID and current ID doesn't match
    if (knownCorrectId && permit.facility_id !== knownCorrectId) {
      console.log(`✗ WRONG FACILITY ID!`);
      console.log(`  Current: ${permit.facility_id}`);
      console.log(`  Should be: ${knownCorrectId}`);

      updatesNeeded.push({
        permitId: permit.id,
        userId: permit.user_id,
        permitName: permit.name,
        oldId: permit.facility_id,
        newId: knownCorrectId,
        reason: 'Matched known facility'
      });

      await sleep(500);
      continue;
    }

    // Test current facility ID
    console.log('Testing current facility ID...');
    const result = await testFacilityId(permit.facility_id);

    if (result.valid) {
      console.log('✓ Facility ID is VALID');
      console.log(`  Recreation.gov Name: ${result.facilityName}`);
      if (result.divisions) {
        const divEntries = Object.entries(result.divisions).slice(0, 5);
        if (divEntries.length > 0) {
          console.log(`  Divisions (${Object.keys(result.divisions).length}):`);
          for (const [divId, divInfo] of divEntries) {
            console.log(`    - ${divId}: ${divInfo.name || 'Unknown'}`);
          }
        } else {
          console.log(`  ⚠️  WARNING: No divisions found - may not have availability data`);
        }
      }
    } else {
      console.log('✗ Facility ID is INVALID');
      console.log(`  Error: ${result.error}`);

      // Try to find correct facility ID
      console.log('\n  Searching for correct facility ID...');

      let correctId = null;

      // Search recreation.gov
      console.log(`  Searching recreation.gov for: ${permit.name}`);
      correctId = await searchFacilityByName(permit.name);
      if (correctId) {
        console.log(`  Found via search: ${correctId}`);
        // Verify it's valid
        const verify = await testFacilityId(correctId);
        if (verify.valid) {
          console.log(`  Verified: ${verify.facilityName}`);
        } else {
          console.log('  Search result was invalid, ignoring');
          correctId = null;
        }
      }

      if (correctId && correctId !== permit.facility_id) {
        updatesNeeded.push({
          permitId: permit.id,
          userId: permit.user_id,
          permitName: permit.name,
          oldId: permit.facility_id,
          newId: correctId,
          reason: 'Found via search'
        });
      } else {
        console.log('  ✗ Could not find correct facility ID');
      }
    }

    await sleep(500); // Be nice to the API
  }

  // Show summary of updates needed
  console.log('\n' + '='.repeat(70));
  console.log('UPDATE SUMMARY');
  console.log('='.repeat(70));

  if (updatesNeeded.length === 0) {
    console.log('\n✓ All facility IDs are valid! No updates needed.');
    return;
  }

  console.log(`\nFound ${updatesNeeded.length} permit(s) that need updating:\n`);
  for (const update of updatesNeeded) {
    console.log(`  Permit #${update.permitId}: ${update.permitName}`);
    console.log(`    ${update.oldId} → ${update.newId}`);
    if (update.reason) {
      console.log(`    Reason: ${update.reason}`);
    }
    console.log();
  }

  // Apply updates automatically
  console.log('Applying updates...');
  for (const update of updatesNeeded) {
    process.stdout.write(`  Updating permit #${update.permitId}... `);
    try {
      updatePermitFacilityId(update.permitId, update.userId, update.newId);
      console.log('✓');
    } catch (error) {
      console.log(`✗ Error: ${error.message}`);
    }
  }

  console.log('\n✓ Updates complete!');
  console.log('\nRecommendation: Run the "Check Permits Now" feature to verify the fixes.');
}

main().catch(error => {
  console.error('Error:', error);
  process.exit(1);
});

File: fix_facility_ids.py
#!/usr/bin/env python3
"""
Script to verify and fix facility IDs for all permit monitors
"""
import sqlite3
import requests
import json
import time
from typing import Dict, Optional, List

DB_PATH = '/home/user/river_permits/database/permits.db'

# Known correct facility IDs based on investigation
KNOWN_FACILITIES = {
    "Gates of Lodore": "250014",  # Dinosaur Green And Yampa River Permits
    "Yampa River": "250014",      # Same facility, different division
    "Deerlodge Park": "250014",   # Same facility, different division
    "San Juan River": "234621",   # San Juan River
    "Desolation Canyon": "233393", # Desolation Gray - Green River Permit
    "Grand Canyon": "233262",     # Grand Canyon River
}

def get_all_permits():
    """Get all permits from the database"""
    conn = sqlite3.connect(DB_PATH)
    conn.row_factory = sqlite3.Row
    cursor = conn.cursor()

    cursor.execute("""
        SELECT p.*, u.email
        FROM permits p
        JOIN users u ON p.user_id = u.id
        ORDER BY p.id
    """)

    permits = [dict(row) for row in cursor.fetchall()]
    conn.close()

    return permits

def test_facility_id(facility_id: str) -> Dict:
    """Test if a facility ID is valid and get its info"""
    url = f"https://www.recreation.gov/api/permits/{facility_id}"
    headers = {
        'User-Agent': 'Mozilla/5.0 (Windows NT 10.0; Win64; x64) AppleWebKit/537.36'
    }

    try:
        response = requests.get(url, headers=headers, timeout=10)
        if response.status_code == 200:
            data = response.json()
            if 'payload' in data:
                payload = data['payload']
                return {
                    'valid': True,
                    'facility_name': payload.get('facility_name', 'Unknown'),
                    'divisions': payload.get('divisions', {}),
                    'data': payload
                }
        return {'valid': False, 'error': f"HTTP {response.status_code}"}
    except Exception as e:
        return {'valid': False, 'error': str(e)}

def search_facility_by_name(name: str) -> Optional[str]:
    """Search for a facility by name on recreation.gov"""
    url = "https://www.recreation.gov/api/search"
    params = {
        "q": name,
        "entity_type": "permit"
    }
    headers = {
        'User-Agent': 'Mozilla/5.0 (Windows NT 10.0; Win64; x64) AppleWebKit/537.36'
    }

    try:
        response = requests.get(url, params=params, headers=headers, timeout=10)
        if response.status_code == 200:
            data = response.json()
            if 'results' in data and len(data['results']) > 0:
                # Return the first result's entity_id
                result = data['results'][0]
                return result.get('entity_id')
    except Exception as e:
        print(f"    Search error: {e}")

    return None

def update_facility_id(permit_id: int, new_facility_id: str):
    """Update a permit's facility ID in the database"""
    conn = sqlite3.connect(DB_PATH)
    cursor = conn.cursor()

    cursor.execute("""
        UPDATE permits
        SET facility_id = ?
        WHERE id = ?
    """, (new_facility_id, permit_id))

    conn.commit()
    conn.close()

def main():
    print("=" * 70)
    print("Facility ID Verification and Fix Script")
    print("=" * 70)

    permits = get_all_permits()

    if not permits:
        print("\nNo permits found in database!")
        return

    print(f"\nFound {len(permits)} permit(s) to check\n")

    updates_needed = []

    for permit in permits:
        print(f"\n{'=' * 70}")
        print(f"Permit #{permit['id']}: {permit['name']}")
        print(f"User: {permit['email']}")
        print(f"Current Facility ID: {permit['facility_id']}")
        print('=' * 70)

        # Test current facility ID
        print("Testing current facility ID...")
        result = test_facility_id(permit['facility_id'])

        if result['valid']:
            print(f"✓ Facility ID is VALID")
            print(f"  Recreation.gov Name: {result['facility_name']}")
            if result['divisions']:
                print(f"  Divisions ({len(result['divisions'])}):")
                for div_id, div_info in list(result['divisions'].items())[:5]:
                    print(f"    - {div_id}: {div_info.get('name', 'Unknown')}")
        else:
            print(f"✗ Facility ID is INVALID")
            print(f"  Error: {result['error']}")

            # Try to find correct facility ID
            print(f"\n  Searching for correct facility ID...")

            # First check known facilities
            correct_id = None
            for known_name, known_id in KNOWN_FACILITIES.items():
                if known_name.lower() in permit['name'].lower():
                    correct_id = known_id
                    print(f"  Found in known facilities: {known_id}")
                    break

            # If not found in known facilities, search recreation.gov
            if not correct_id:
                print(f"  Searching recreation.gov for: {permit['name']}")
                correct_id = search_facility_by_name(permit['name'])
                if correct_id:
                    print(f"  Found via search: {correct_id}")
                    # Verify it's valid
                    verify = test_facility_id(correct_id)
                    if verify['valid']:
                        print(f"  Verified: {verify['facility_name']}")
                    else:
                        print(f"  Search result was invalid, ignoring")
                        correct_id = None

            if correct_id and correct_id != permit['facility_id']:
                updates_needed.append({
                    'permit_id': permit['id'],
                    'permit_name': permit['name'],
                    'old_id': permit['facility_id'],
                    'new_id': correct_id
                })
            else:
                print(f"  ✗ Could not find correct facility ID")

        time.sleep(0.5)  # Be nice to the API

    # Show summary of updates needed
    print("\n" + "=" * 70)
    print("UPDATE SUMMARY")
    print("=" * 70)

    if not updates_needed:
        print("\n✓ All facility IDs are valid! No updates needed.")
        return

    print(f"\nFound {len(updates_needed)} permit(s) that need updating:\n")
    for update in updates_needed:
        print(f"  Permit #{update['permit_id']}: {update['permit_name']}")
        print(f"    {update['old_id']} → {update['new_id']}")
        print()

    # Ask for confirmation
    response = input("\nApply these updates? (yes/no): ").strip().lower()

    if response in ['yes', 'y']:
        print("\nApplying updates...")
        for update in updates_needed:
            print(f"  Updating permit #{update['permit_id']}...", end=' ')
            try:
                update_facility_id(update['permit_id'], update['new_id'])
                print("✓")
            except Exception as e:
                print(f"✗ Error: {e}")

        print("\n✓ Updates complete!")
        print("\nRecommendation: Run the 'Check Permits Now' feature to verify the fixes.")
    else:
        print("\nNo updates applied.")

if __name__ == "__main__":
    main()
